File: grading_scripts/pa1/List.java
public class List{
	
   //Private Node class for doubly linked list
   private class node{
      int value;
      node next;
      node prev;
	  node(){}
	  node(int value){
		  this.value = value;
	  }
   }
   
   //List Fields
   int size = 0;
   node cursor = null; 
   node front = new node();
   node back = new node();
	
   // Constructor
   List(){	   // Creates a new empty list.
	   front.next = back;
	   front.prev = null;
	   back.next = null;
	   back.prev = front;
	   size = 0;
   }
   
   // Access functions
   int length(){ // Returns the number of elements in this List.
       return size;
   }
   int index(){ // If cursor is defined, returns the index of the cursor element, otherwise returns -1.
       if(cursor == null){
		   return -1;
	   }
       node iter = front;
	   int index_counter = 0;
	   while(iter != cursor){
		   index_counter++;
		   iter = iter.next;
	   }
	   return index_counter;
   }
   int front(){ // Returns front element. Pre: length()>0
       return front.value;
   }
   int back(){ // Returns back element. Pre: length()>0
       return back.value;
   }
   int get(){ // Returns cursor element. Pre: length()>0, index()>=0
	   return cursor.value;
   }
   boolean equals(List L) {			// Returns true if and only if this List and L are the same integer sequence. 
		if (L.length() != size){	// The states of the cursors in the two Lists are not used in determining equality.
			return false;
		}
		node holder = cursor;	
		node listholder = L.cursor;
        L.cursor = L.front;      
		cursor = front;
		while(L.cursor != null){
			if(L.get() != cursor.value){
				cursor = holder;
				L.cursor = listholder;
				return false;
			}
			L.cursor = L.cursor.next;
			cursor = cursor.next;
		}
		return true;
   }
 
  // Manipulation procedures
  void clear(){ // Resets this List to its original empty state.
		front.value = 0;
		back.value = 0;
		front.next = back;
		front.prev = null;
		back.next = null;
		back.prev = front;
		size = 0;
                cursor = null;
  }
  void moveFront(){ 	// If List is non-empty, places the cursor under the front element,
	   if(size != 0){	// otherwise does nothing.
		   cursor = front;
	   }
  }
  void moveBack(){ 		// If List is non-empty, places the cursor under the back element,
		if(size != 0){	// otherwise does nothing.
		   cursor = back;
	   }
  }
  void movePrev(){ 					// If cursor is defined and not at front, moves cursor one step toward
		if(cursor != null){			// front of this List, if cursor is defined and at front, cursor becomes
			cursor = cursor.prev;	// undefined, if cursor is undefined does nothing.
		}
  }
  void moveNext(){ 					// If cursor is defined and not at back, moves cursor one step toward
		if(cursor != null){			// back of this List, if cursor is defined and at back, cursor becomes
			cursor = cursor.next;	// undefined, if cursor is undefined does nothing.
		}
  }
  void prepend(int data){ 			// Insert new element into this List. If List is non-empty,
		if(size == 0){				// insertion takes place before front element.
			node new_element = new node(data);
			front = new_element;
			back = new_element;
			size++;
		}else
		if(size == 1){
			node new_element = new node(data);
			new_element.next = back;
			back.prev = new_element;
			front = new_element;
			size++;
		}else{
			node new_element = new node(data);
			front.prev = new_element;
			new_element.next = front;
			new_element.prev = null;
			front = new_element;
			size++;
		}
		
  }
  void append(int data){ 		// Insert new element into this List. If List is non-empty,
		if(size == 0){			// insertion takes place after back element.
			node new_element = new node(data);
			front = new_element;
			back = new_element;
			size++;
		}else
		if(size == 1){
			node new_element = new node(data);
			new_element.prev = front;
			front.next=new_element;
			back = new_element;
			size++;
		}else{
			node new_element = new node(data);
			back.next = new_element;
			new_element.next = null;
			new_element.prev = back;
			back = new_element;
			size++;
		}
  }
  void insertBefore(int data){ 							// Insert new element before cursor.
		if(cursor == front || size == 1){				// Pre: length()>0, index()>=0
			prepend(data);
		}else{
			node new_element = new node(data);
			new_element.next = cursor;
			new_element.prev = cursor.prev;
			cursor.prev.next = new_element;
			cursor.prev = new_element;
			size++;
		}
  }
  void insertAfter(int data){ 				  // Inserts new element after cursor.
		if(cursor == back || size == 1){	  // Pre: length()>0, index()>=0
			append(data);
		}else{
			node new_element = new node(data);
			new_element.next = cursor.next;
			new_element.prev = cursor;
			cursor.next.prev = new_element;
			cursor.next = new_element;
			size++;
		}
  }
  void deleteFront(){ 		// Deletes the front element. Pre: length()>0
		if(cursor == front){// Pre: length()>0, index()>=0
			cursor = null;
		}
		if(size == 1){		
			front.value = 0;
			back.value = 0;
			front.next = back;
			front.prev = null;
			back.next = null;
			back.prev = front;
			size = 0;	
		}else{
			front = front.next;	
			size--;
		}
  }
  void deleteBack(){ 		// Deletes the back element. Pre: length()>0
		if(cursor == back){	// Pre: length()>0, index()>=0
			cursor = null;
		}
		if(size == 1){		
			front.value = 0;
			back.value = 0;
			front.next = back;
			front.prev = null;
			back.next = null;
			back.prev = front;
			size = 0;
		}else{
			back = back.prev;	
			back.next = null;
			size--;
		}
  }
  void delete(){ 			// Deletes cursor element, making cursor undefined.
	   if(cursor == back){
		   deleteBack();
		   cursor = null;
	   }
	   else if(cursor == front){
		   deleteFront();
		   cursor = null;
	   }else{
			cursor.prev.next = cursor.next;
			cursor.next.prev = cursor.prev;
			cursor = null;
			size--;
	   }
  }
	
  // Other methods
  public String toString(){ 		// Overrides Object's toString method. Returns a String
		String sequence = "";		// representation of this List consisting of a space
		node iter = front;			// separated sequence of integers, with front on left.
		for(int i = 0; i < size; i++){
			sequence += Integer.toString(iter.value);
			if(i + 1 != size){
				sequence += " ";
			}
			iter = iter.next;
		}
		return sequence;
  }
  
  List copy(){ 						// Returns a new List representing the same integer sequence as this
	  List copy = new List();		// List. The cursor in the new list is undefined, regardless of the
	  node iter = front;			// state of the cursor in this List. This List is unchanged.
	  if(size == 1){
		  copy.front = front;
		  copy.back = back;
		  return copy;
	  }
	  for(int i = 0;i<size;i++){
			copy.append(iter.value);
			iter = iter.next;
			
	  }
	  return copy;
  } 
  
  
}

File: pa5/List.c
#include<stdio.h>
#include<stdlib.h>
#include "List.h"

typedef struct NodeObj{
        int number;
        struct NodeObj* next;
        struct NodeObj* prev;
}NodeObj;

typedef NodeObj* Node;

typedef struct ListObj{
        int size;
        Node cursor;
        Node front;
        Node back;
}ListObj;

// Constructors-Destructors ---------------------------------------------------
List newList(void){
   List newlist = malloc(sizeof(ListObj));
   newlist->cursor = NULL;
   newlist->front = NULL;
   newlist->back = NULL;
   newlist->size = 0;
   return newlist;
}
void freeList(List* pL){
   clear(*pL);
   free(*pL); 
}

// Access functions -----------------------------------------------------------
int length(List L){
   return L->size;
}
int index(List L){
   if(L->cursor == NULL){
      return -1;
   }else{
      int index = 0;
      Node iter = L->front;
      while(iter != L->cursor){
         index++;
         iter = iter->next;
      }
      return index;
   }
}
int front(List L){
   return L->front->number;
}
int back(List L){
   return L->back->number;
}
int get(List L){
   return L->cursor->number;
}
int equals(List A, List B){
   if(A->size != B->size){
      return 0;
   }
   Node A_iter = A->front;
   Node B_iter = B->front;
   for(int i = 0; i < A->size; i++){
      if(A_iter->number != B_iter->number){
         return 0;
      }
      A_iter = A_iter->next;
      B_iter = B_iter->next;
   }
   return 1;
}

// Manipulation procedures ----------------------------------------------------
void clear(List L){
   if(L->size == 1){
      Node deleted = L->front;
      free(deleted);
      L->front = NULL;
      L->back = NULL;
      L->cursor = NULL;
      L->size = 0;
   }else{
      //For some reason, looping deleteFront gives no errors 
      //vs. iterating and freeing
      int size = L->size;
      for(int i = 0; i < size; i++){
         deleteFront(L);
      }
      L->cursor = NULL;
   }
}
void moveFront(List L){
   if(L->size != 0){
      L->cursor = L->front;
   }
}
void moveBack(List L){
   if(L->size != 0){
      L->cursor = L->back;
   }
}
void movePrev(List L){
   if(L->cursor != NULL){
      L->cursor = L->cursor->prev;
   }
}
void moveNext(List L){
   if(L->cursor != NULL){
      L->cursor = L->cursor->next;
   }
}
void prepend(List L, int data){
   if(length(L) == 0){
      Node newelement = malloc(sizeof(NodeObj));
      newelement->number = data;
      L->front = newelement;
      L->back = newelement;
      L->size++;
   }else if(length(L) == 1){
      Node newelement = malloc(sizeof(NodeObj));
      newelement->number = data;
      newelement->next = L->back;
      L->back->prev = newelement;
      L->front = newelement;
      L->size++;
   }else{
      Node newelement = malloc(sizeof(NodeObj));
      newelement->number = data;
      newelement->next = L->front;
      L->front->prev = newelement;
      L->front = newelement;
      L->size++;
   }
}
void append(List L, int data){
   if(length(L) == 0){
      Node newelement = malloc(sizeof(NodeObj));
      newelement->number = data;
      L->front = newelement;
      L->back = newelement;
      L->size++;
   }else if (length(L) == 1){
      Node newelement = malloc(sizeof(NodeObj));
      newelement->number = data;
      newelement->prev = L->front;
      L->front->next = newelement;
      L->back = newelement;
      L->size++;
   }else{
      Node newelement = malloc(sizeof(NodeObj));
      newelement->number = data;
      newelement->prev = L->back;
      L->back->next = newelement;
      L->back = newelement;
      L->size++;
   }
}
void insertBefore(List L, int data){
   if (L->cursor == L->front || L->size == 1){
      prepend(L,data);
   }else{
      Node newelement = malloc(sizeof(NodeObj));
      newelement->number = data;
      newelement->next = L->cursor;
      newelement->prev = L->cursor->prev;
      L->cursor->prev->next = newelement;
      L->cursor->prev = newelement;
      L->size++;
   }
}
void insertAfter(List L, int data){
   if(L->cursor == L->back || L->size == 1){
      append(L,data);
   }else{
      Node newelement = malloc(sizeof(NodeObj));
      newelement->number = data;
      newelement->next = L->cursor->next;
      newelement->prev = L->cursor;
      L->cursor->next->prev = newelement;
      L->cursor->next = newelement;
      L->size++;
   }
}
void deleteFront(List L){
   if(L->cursor == L->front){
      L->cursor = NULL;
   }
   if(L->size == 1){
      free(L->front);
      L->front = NULL;
      L->back = NULL;
      L->size--;
   }else{
      Node deleted = L->front;
      L->front = L->front->next;
      free(deleted);
      L->size--;
   }
}
void deleteBack(List L){
   if(L->cursor == L->back){
      L->cursor = NULL;
   }
   if(L->size == 1){
      free(L->back);
      L->front = NULL;
      L->back = NULL;
      L->size--;
   }else{
      Node deleted = L->back;
      L->back = L->back->prev;
      free(deleted);
      L->size--;
   }
}
void delete(List L){
   if(L->cursor == L->front){
      deleteFront(L);
      L->cursor = NULL;
   }else if(L->cursor == L->back){
      deleteBack(L);
      L->cursor = NULL;
   }else{
      L->cursor->prev->next = L->cursor->next;
      L->cursor->next->prev = L->cursor->prev;
      Node deleted = L->cursor;
      free(deleted);
      L->cursor = NULL;
      L->size--;
   }
}

// Other operations -----------------------------------------------------------
void printList(FILE* out, List L){
   Node iter = L->front;
   for(int i = 0; i < L->size; i++){
      fprintf(out,"%d", iter->number);
      iter = iter->next;
      if(i + 1 != L->size){
         fprintf(out,"%s"," ");
      }
   }
}
List copyList(List L){
   List copy = newList();
   Node iter = L->front;
   if(L->size == 1){
      copy->front = L->front;
      copy->back = L->back;
      copy->size++;
   }else{
      for(int i = 0; i < L->size; i++){
         append(copy, iter->number);
         iter = iter->next;
      }
   }
   return copy;
}
int isFront(List L){
   if(L->cursor == L->front){
      return 1;
   }
   return 0;
}
int isBack(List L){
   if(L->cursor == L->back){
      return 1;
   }
   return 0;
}
